community/values/src/main/java/org/neo4j/values/storable/StringValue.java

```diff
@@ -55,7 +52,56 @@ public boolean equals( String x )
     @Override
     public int computeHash()
     {
-        return value().hashCode();
+        //NOTE that we are basing the hash code on code points instead of char[] values.
+        String value = value();
+        if ( value.isEmpty() )
+        {
+            return 0;
+        }
+        int h = 1;
+        final int length = value.length();
+        for ( int offset = 0; offset < length; )
+        {
+            final int codePoint = value.codePointAt( offset );
+            h = 31 * h + codePoint;
+            offset += Character.charCount( codePoint );
```

Comment:
I typically put the increment in the header of the loop. It requires declaring the `codePoint` variable there too, but I think it is worth it for the clarity of having the progression of the loop declared in the header.

Reply:
Do you mean like

```
 for ( int offset = 0, codePoint = value.codePointAt( offset ); offset < length;codePoint = value.codePointAt( offset ), offset += Character.charCount( codePoint ) )
{
  h = 31 * h + codePoint;
}
```
looks a little bit weird?

Reply:
No, like this:

    for ( int offset = 0, codePoint; offset < length; offset += Character.charCount( codePoint ) )
    {
        codePoint = value.codePointAt( offset );
        h = 31 * h + codePoint;
    }